server/src/main/java/io/druid/client/CachingClusteredClient.java

```diff
@@ -444,12 +446,19 @@ public void run()
                                   {
                                     if (cachePopulator != null) {
                                       try {
-                                        Futures.allAsList(cacheFutures).get();
+                                        // For some reason, this can get ahead of the evaluation of the sequence that
```

Comment:
Instead of waiting on the future with an arbitrary timeout, how about we add a callback on the future to do the cache population in the backgroundExecutorService, e.g. `Futures.addCallback(Futures.allAsList(cacheFutures), { ... cache populating callback ... }, backgroundExecutorService)`.
Then we can also execute the `withEffect` in the same thread as the one returning the sequence and avoid any nasty execution ordering problems.